src/commonMain/kotlin/com.adamratzman.spotify/SpotifyRestAction.kt

```diff
@@ -0,0 +1,196 @@
+/* Spotify Web API - Kotlin Wrapper; MIT License, 2019; Original author: Adam Ratzman */
+package com.adamratzman.spotify
+
+import com.adamratzman.spotify.models.AbstractPagingObject
+import com.adamratzman.spotify.utils.TimeUnit
+import com.adamratzman.spotify.utils.getCurrentTimeMs
+import com.adamratzman.spotify.utils.schedule
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.emitAll
+import kotlinx.coroutines.flow.flatMapConcat
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Provides a uniform interface to retrieve, whether synchronously or asynchronously, [T] from Spotify
+ */
+open class SpotifyRestAction<T> internal constructor(protected val api: SpotifyApi, val supplier: () -> T) {
+    private var hasRunBacking: Boolean = false
+    private var hasCompletedBacking: Boolean = false
+
+    /**
+     * Whether this REST action has been *commenced*.
+     *
+     * Not to be confused with [hasCompleted]
+     */
+    fun hasRun() = hasRunBacking
+
+    /**
+     * Whether this REST action has been fully *completed*
+     */
+    fun hasCompleted() = hasCompletedBacking
+
+    /**
+     * Invoke [supplier] and synchronously retrieve [T]
+     */
+    fun complete(): T {
+        hasRunBacking = true
+        return try {
+            supplier().also { hasCompletedBacking = true }
+        } catch (e: Throwable) {
+            throw e
+        }
```

Comment:
Why is this wrapped in a `try/catch`? This doesn't change anything.

Reply:
Good question. Thanks for catching these idiotic changes 😄